packages/buildcop/buildcop.go

```diff
@@ -0,0 +1,208 @@
+
+	flag.Parse()
+
+	log.Println("Sending logs to Build Cop Bot...")
+	if ok := publish(*projectID, *topicID, *repo, *installationID); !ok {
+		os.Exit(1)
+	}
+	log.Println("Done!")
+}
+
+type githubInstallation struct {
+	ID string `json:"id"`
+}
+
+type message struct {
+	Name         string
+	Type         string
+	Location     string
+	installation githubInstallation
+	repo         string
+	buildID      string
+	buildURL     string
+	xunitXML     string
+}
+
+// publish searches for sponge_log.xml files and publishes them to Pub/Sub.
+// publish logs a message and returns false if there was an error.
+func publish(projectID, topicID, repo, installationID string) (ok bool) {
+	ctx := context.Background()
+
+	gfileDir := os.Getenv("KOKORO_GFILE_DIR")
+	if gfileDir == "" {
+		log.Println("KOKORO_GFILE_DIR not set, unable to get service account")
+		return false
+	}
+	saPath := filepath.Join(gfileDir, "kokoro-trampoline.service-account.json")
+
+	client, err := pubsub.NewClient(ctx, projectID, option.WithCredentialsFile(saPath))
+	if err != nil {
+		log.Printf("Unable to connect to Pub/Sub: %v", err)
+		return false
+	}
+	topic := client.Topic(topicID)
+
+	if repo == "" {
+		var ok bool
+		repo, ok = detectRepo()
+		if !ok {
+			log.Print(`Unable to detect repo. Please set the --repo flag.
+If your repo is github.com/GoogleCloudPlatform/golang-samples, --repo should be GoogleCloudPlatform/golang-samples.
+
+If your repo is not in GoogleCloudPlatform or googleapis, you must also set
+--installation_id. See https://github.com/apps/build-cop-bot/.`)
+			return false
+		}
+	}
+
+	if installationID == "" {
+		var ok bool
+		installationID, ok = detectInstallationID(repo)
+		if !ok {
+			log.Print(`Unable to detect installation ID. Please set the --installation_id flag.
+If your repo is part of GoogleCloudPlatform or googleapis and you see this error,
+file an issue at https://github.com/googleapis/repo-automation-bots/issues.
+Otherwise, set --installation_id with the numeric installation ID.
+See https://github.com/apps/build-cop-bot/.`)
+			return false
+		}
+	}
+
+	// Handle logs in the current directory.
+	if err := filepath.Walk(".", processLog(ctx, repo, installationID, topic)); err != nil {
+		log.Printf("Error publishing logs: %v", err)
+		return false
+	}
+
+	// Handle logs in the KOKORO_ARTIFACTS_DIR directory.
+	artifactsDir := os.Getenv("KOKORO_ARTIFACTS_DIR")
+	currentPath, err := filepath.Abs(".")
+	if err != nil {
+		// If there is an error getting the current path, don't error out.
+		log.Printf("Unable to get current directory: %v: continuing...", err)
+		return true
+	}
+	if artifactsDir != "" && artifactsDir != currentPath {
+		if err := filepath.Walk(artifactsDir, processLog(ctx, repo, installationID, topic)); err != nil {
+			log.Printf("Error publishing logs: %v", err)
+			return false
+		}
+	}
+	return true
+}
+
+// detectRepo tries to detect the repo from the environment.
+func detectRepo() (repo string, ok bool) {
```

Comment:
I don't think "" has special meaning for these, do you need to return a bool as well?

Reply:
Removed the bool.